test_util/helpers.py

```diff
@@ -23,81 +24,137 @@ def path_join(p1, p2):
+            query if query is not None else self.query,
+            fragment if fragment is not None else self.fragment,
+            port if port is not None else self.port)
+
 
-    def __init__(self, default_host_url, api_base, default_headers=None,
-                 ca_cert_path=None, get_node_url=None):
-        self.default_host_url = default_host_url
-        self.api_base = api_base
-        if default_headers is None:
-            default_headers = dict()
-        self.default_headers = default_headers
-        self.ca_cert_path = ca_cert_path
-        self._get_node_url = get_node_url
-
-    def api_request(self, method, path, host_url=None, node=None, port=None, **kwargs):
+class ApiClient:
+    """utilizes requests.session with some URL handling to remove boilerplate for making
+    cluster requests. Can also be used to bind advanced helper methods of a given service
+    """
+    def __init__(self, url: Url, get_node_port=None):
         """
-        Makes a request with default headers + user auth headers if necessary
-        If self.ca_cert_path is set, this method will pass it to requests
         Args:
-            method: (str) name of method for requests.request
-            path: see get_url()
-            host_url: override the client's default host URL
-            node: if a get_node_url function is set, node and port will be passed to this function
-                instead of setting a host_url or using the default
-            port: can be used with node with get_node_url is set
-            **kwargs: any optional arguments to be passed to requests.request
+            url: Url object to which requests will be made
+        Keyword Args:
+            get_node_port: a callback that takes a node string as an argument. This function
+                must return the port for that string. Intended for communicating with an API
+                that uses different ports on different hosts (e.g. AdminRouter, Mesos )
         """
-        headers = copy.copy(self.default_headers)
+        self.url = url
+        self.session = requests.Session()
+        self._get_node_port = get_node_port
 
-        # allow kwarg to override verification so client can be used generically
-        if self.ca_cert_path and 'verify' not in kwargs:
-            kwargs['verify'] = self.ca_cert_path
+    def get_client(self, scheme=None, host=None, path=None, query=None, fragment=None, port=None):
+        """Takes the same arguments as Url.get_url
+        """
+        clone = copy.deepcopy(self)
+        clone.url = self.url.get_url(scheme=scheme, host=host, path=path, query=query, fragment=fragment, port=port)
+        return clone
 
-        if self.api_base:
-            path = path_join(self.api_base, path)
+    def api_request(self, method, path_ex, scheme=None, host=None, query=None,
+                    fragment=None, path=None, port=None, node=None, **kwargs):
+        """ Direct wrapper for session.request. Returns request.Response
```

Comment:
I am sure you checked this yourself, but for the record: there is no conflict between these many keyword argument names and those consumed by `requests.request` (http://docs.python-requests.org/en/master/api/#requests.request).

That's obviously important, might be worth a comment.